### DISCUSSION TITLE
Uniqueness of NodeIds

### DISCUSSION THREAD

**Discussion by tobias-neubert** [eclipse-milo/milo#892] [General]
Hi all, I hope this is the right place to ask such questions. I think it is an easy one: Does a NodeId have to be unique accross  the whole tree of nodes?

I want to publish a set of values that are leaves of a tree of components, where each component provides the same values. Like so:

```
pool
|-- system1
  | -- component 1
    | -- value1
    | -- value2
    | -- ...
  | -- component 2
    | -- value1
    | -- value 2
```            

I create a namespace containing that tree where the labels that you can see are the (local) ids of the nodes. That works and I can browse the tree in my test.

But if I want to write a value to a specific value node, I try to fetch that node from the ```UaNodeManager.getNode()```, which saves alle nodes in a map keyed by their NodeIds. And in there there are only the last registered values nodes of my tree. So I assume, the NodeIds must be unique.

Am I right?

**Comment by tobias-neubert** [comment]
The server example seems to support my assumption: It prefixes all NodeIds with the parent id string.

**Comment by kevinherron** [comment]
Yes, NodeIds must be unique across the entire server.

**Comment by tobias-neubert** [comment]
Great thank you. Do you know if this is an implementation detail for milo or does the standard defines it this way? I am not sure. The only thing I found in https://reference.opcfoundation.org/Core/docs/Part3/5.2.2/ is this sentence: "Nodes are unambiguously identified using a constructed identifier called the NodeId."

**Comment by kevinherron** [comment]
It’s part of the standard, but I can’t readily find anything more explicit than the passage you found.
